### ISSUE TITLE
Support data mocks of arbitrary variables inside a Rego policy

### ISSUE THREAD

**Issue by artis3n** [open-policy-agent/opa#3639]
## Expected Behavior

Similar behavior as <https://www.openpolicyagent.org/docs/latest/policy-testing/#data-mocking> to allow to mock arbitrary variables or function outputs for policy scenarios with branching conditions.

## Actual Behavior

Only `input` and `data.*` variables can be mocked, making it harder to create unit tests with policies that include [http.send](https://www.openpolicyagent.org/docs/latest/policy-reference/#http) calls and make a branching decision based on that output.

## Additional Info

I have a Rego policy that uses `http.send` to make a request to pull down the latest version of a JSON document. It then parses the result and has a snippet similar to the following:

```rego
outer_function_call = result {
  output := http.send({...})

  result := determine_a_or_b(output)
}

# Assume a and b exist for the purpose of this example
determine_a_or_b(output) = a {
  not output.expectation
} else = b {
  true
}
```

I would like to be able to mock `output` such that we can cover the else condition in `determine_a_or_b`, as currently the output of the HTTP call does not hit that line. It would in several months as the data source we're calling updates, at which point we'll always hit the `else` block and miss coverage of the initial condition. Alternatively, some way to mock the response for an HTTP call would solve this nicely.

Because the input comes from the HTTP call and not from a `data` input, I can't mock the data inside a unit test to get coverage on the else condition in `determine_a_or_b`. We have a CI check to ensure our policies have full coverage so that we don't push out a policy with broken logic to the org. We use the make command:

```make
.PHONY: coverage
coverage:
	if [ ! $$(opa test --coverage policies | jq '.coverage') == "100" ]; then echo "\nThis repository requires 100% coverage on policy code. You are missing coverage: $$(opa test --coverage policies | jq '.coverage')%." && exit 1; fi;
```

Discussions of what is a proper amount of test coverage notwithstanding, we'd prefer to have confidence about the policies we push out to engineering via 100% coverage. We can lower the coverage percentage requirement to handle this edge case that we're having difficulty testing for but that's less "nice."

I am creating another issue to allow for excluding certain lines or functions from `--coverage` reports, https://github.com/open-policy-agent/opa/issues/3640.

**Comment by ashutosh-narkar** [comment]
We have an open issue https://github.com/open-policy-agent/opa/issues/2028 related to adding support to mock functions. 

Currently although you cannot directly mock out `http.send`,  you could put it inside a rule and mock that out. For example something like [this](https://play.openpolicyagent.org/p/t2DhxMbkKl).

**Comment by artis3n** [comment]
Ah yup, #2028 looks like what I'm asking for hehe. This is a duplicate of that issue.

**Comment by artis3n** [comment]
When I tried to mock a function in the way you describe I received the same error, but let me look into that further, that would work as an alternative for the time being

**Comment by artis3n** [comment]
Ah ok, so formatting my function like this:

```rego
my_function = result {
  # ...
}
```

And then mocking it in the test as

```rego
count(warn) == 0 with input as {...} with my_function as { "MOCKED": true }
```

Successfully shows the test evaluating the `result` variable with the `MOCKED` value in it. That unblocks my current need. Will close this ticket in favor of watching #2028 for mocking function calls with input parameters.
